### DISCUSSION TITLE
Force a rebuild of all my packages on r-universe?

### DISCUSSION THREAD

**Discussion by HenrikBengtsson** [r-universe-org/help#140] [Q&A] [answered]
Hi,

how can I force a rebuild of all my packages on r-universe?  

Other that I don't want to spend the time to do it manually for 15-20 packages, I also don't want to push to the tracked branch, because they are meant to be frozen until the next CRAN release.

# Background

I updated my https://github.com/HenrikBengtsson/universe/blob/develop/packages.json, to use:

```json
    "branch": "master",
```

for all packages, because my default branch is `develop` and I do not won't users to install those my mistake.

I pushed this, which triggered 'ropensci-bot' to start rebuilding all the packages.

However, it looks like packages several packages failed to build, e.g. https://henrikbengtsson.r-universe.dev/ui#builds.  Note that all these packages and version pass the CRAN checks and my GitHub Action checks.  So, I suspect there's something not working on r-universe.

Looking at the GHA check logs, it looks like the packages that fail are checked with package versions previously on r-universe (my 'develop' version), not what's available on CRAN (same as my 'master' version). It might also have to do with in what order 'r-universe' was rebuilding the packages. This means that some packages fail because they depend on features that have been pulled from the 'develop' versions. There might be other things going on too, but it's quite nested to track down.

So, I'm hoping a rebuild of all my packages on r-universe might be able to resolve this.

**Comment by jeroen** [comment, marked as answer]
Thanks for the suggestion. I have added a new API for this. You can trigger it using the circle-arrow icon on https://henrikbengtsson.r-universe.dev/

This is very experimental, please let me know if it works for you!

**Reply by HenrikBengtsson** [reply]
Brilliant and, wow, fast! This seems like a nice solution. One comment, though, which I'm sure you're already aware of: Anyone in the world can trigger a rebuild by click those icons - on purpose or not.

**Comment by jeroen** [comment]
Yes I know, for now there is no authentication so anyone can click the button indeed. To prevent abuse I have capped them at 1 attempt per hour per package and you can only trigger rebuilds for runs there were previously unsuccessful.

**Reply by HenrikBengtsson** [reply]
Sounds like a sensible solution.

**Reply by ralmond** [reply]
Thanks for the retry button.  Very useful and I have some packages which are not compiling on the server and it is helpful to tweak them and test to see if it fixes the setup on the server.

1) The one retry/hour limit seems a bit stiff for debugging purposes, as I may need to try many things in a row.  If it is necessary for server capacity issues, that is reasonable, but as an anti-spam measure, it is not ideal.
2) Probably the retry button should check to make sure there has been a new commit to the appropriate branch in the github repository.  Thus, you can't just keep pressing the button, you need to make a change first.  (This could be a more reasonable spam test, as github commits do require authentication.)
3) I feel like there is a race condition between pressing the retry button and pushing/pulling from github.  In particular, the last test I ran looks like it is using the old version of Makefile.in on my project.

**Reply by jeroen** [reply]
@ralmond you are confusing two steps in r-universe. We sync [your monorepo](https://github.com/r-universe/ralmond) with upstream git repositories once per hour. That does not change, we do at most one update per hour per package.

The rebuild button is specifically retry the latest build. So it does not check again if there are upstream changes, it just tries to rebuild the version that we tried last time.

**Comment by ralmond** [comment]
Thanks, that is a quite helpful clarification as I'm currently trying to debug some issues with packages building on my local machine and not on the r-universe build machines.

So my strategy should be fix and test locally, push to github and then check back in an hour to see if the rebuild succeeded.
